blocks/filter/include/gnuradio-4.0/filter/time_domain_filter.hpp

```diff
@@ -107,9 +116,151 @@ a are the feedback coefficients
+            if (!result.has_value()) {
+                throw gr::exception(fmt::format("invalid value for {}: {}", magic_enum::enum_type_name<V>(), str));
+            }
+            return result.value();
+        };
+
+        _filter_type       = cast_enum(_filter_type, filter_type);
+        _filter_response   = cast_enum(_filter_response, filter_response);
+        _iir_design_method = cast_enum(_iir_design_method, iir_design_method);
+        _fir_design_method = cast_enum(_fir_design_method, fir_design_method);
+
+        // Set up filter parameters
+        FilterParameters params;
+        params.order = filter_order;
+        params.fLow  = f_low;
+        params.fHigh = f_high;
+        params.fs    = sample_rate;
+
+        if (_filter_type == FilterType::FIR) { // design FIR filter
+            _filter = FilterImpl(std::move(fir::designFilter<ValueType>(_filter_response, params, _fir_design_method)));
+        } else if (_filter_type == FilterType::IIR) { // design IIR filter
+            _filter = FilterImpl(iir::designFilter<ValueType>(_filter_response, params, _iir_design_method));
+        }
+    }
+
+    [[nodiscard]] T processOne(T input) noexcept
+    requires(TParent::ResamplingControl::kIsConst)
+    {
+        return _filter.processOne(input);
+    }
+
+    [[nodiscard]] work::Status processBulk(std::span<const T> input, std::span<T> output) noexcept
+    requires(not TParent::ResamplingControl::kIsConst)
+    {
+        assert(output.size() >= input.size() / decimate);
+
+        std::size_t out_sample_idx = 0;
+        for (std::size_t i = 0; i < input.size(); ++i) {
+            T output_sample = _filter.processOne(input[i]);
+
+            if (i % decimate == 0) {
+                output[out_sample_idx++] = output_sample;
+            }
+        }
+        return work::Status::OK;
+    }
+};
+
+template<typename T>
+using BasicFilter = BasicFilterProto<T>;
+
+template<typename T>
+using BasicDecimatingFilter = BasicFilterProto<T, Resampling<1UZ, 1UZ, false>>;
+
+template<typename T>
+struct Decimator : Block<Decimator<T>, Resampling<1UZ, 1UZ, false>> {
+    using TParent     = Block<Decimator<T>, Resampling<1UZ, 1UZ, false>>;
+    using Description = Doc<R""(@brief Basic Decimator Block
+
+This block implements a decimator for downsampling (dropping) input data by a
+configurable factor. Filtering is not included in this implementation so expect
+aliasing and sub-sampling related effects.
+)"">;
+
+    PortIn<T>  in;
+    PortOut<T> out;
+
+    Annotated<gr::Size_t, "decimation factor", Doc<"Factor by which to downsample/drop input data">, Visible> decim{1};
+
+    GR_MAKE_REFLECTABLE(Decimator, in, out, decim);
+
+    void settingsChanged(const property_map& /*oldSettings*/, const property_map& /*newSettings*/) { this->input_chunk_size = decim; }
+
+    [[nodiscard]] work::Status processBulk(std::span<const T> input, std::span<T> output) noexcept {
+        assert(output.size() >= input.size() / decim);
```

Comment:
I believe that it as guaranteed in Block implementation: `n_inputs = k * input_chunk_size`  and `n_outputs = k * output_chunk_size`